File: sdks/typescript/tests/regression.test.ts
/**
 * Comprehensive Regression Test Suite for TypeScript SDK
 *
 * This test suite covers:
 * 1. Basic RDF triple operations (CRUD)
 * 2. SPARQL query functionality
 * 3. Node types (IRI, literal, typed literals, language tags)
 * 4. Hypergraph operations (binary, ternary, n-ary edges)
 * 5. Error handling and edge cases
 * 6. Performance with large datasets
 *
 * Matches Rust SDK test coverage for cross-language consistency.
 */

import { describe, test, expect, beforeEach } from '@jest/globals';
import { GraphDB, Node } from '../src';

describe('Basic CRUD Operations', () => {
    let db: GraphDB;

    beforeEach(() => {
        db = GraphDB.inMemory();
    });

    test('creates empty database', () => {
        expect(db.isEmpty()).toBe(true);
        expect(db.count()).toBe(0);
    });

    test('inserts single triple', async () => {
        await db.insert()
            .triple(
                Node.iri('http://example.org/alice'),
                Node.iri('http://www.w3.org/1999/02/22-rdf-syntax-ns#type'),
                Node.iri('http://xmlns.com/foaf/0.1/Person')
            )
            .execute();

        expect(db.count()).toBe(1);
        expect(db.isEmpty()).toBe(false);
    });

    test('inserts multiple triples in single operation', async () => {
        await db.insert()
            .triple(
                Node.iri('http://example.org/alice'),
                Node.iri('http://xmlns.com/foaf/0.1/name'),
                Node.literal('Alice')
            )
            .triple(
                Node.iri('http://example.org/alice'),
                Node.iri('http://xmlns.com/foaf/0.1/age'),
                Node.integer(30)
            )
            .execute();

        expect(db.count()).toBe(2);
    });

    test('queries all triples', async () => {
        await db.insert()
            .triple(
                Node.iri('http://example.org/test'),
                Node.iri('http://www.w3.org/1999/02/22-rdf-syntax-ns#type'),
                Node.iri('http://example.org/TestClass')
            )
            .execute();

        const results = await db.query()
            .sparql('SELECT ?s ?p ?o WHERE { ?s ?p ?o }')
            .execute();

        expect(results.length).toBe(1);
        expect(results.isEmpty()).toBe(false);
    });

    test('clears all triples', async () => {
        await db.insert()
            .triple(
                Node.iri('http://example.org/test'),
                Node.iri('http://example.org/p'),
                Node.literal('value')
            )
            .execute();

        expect(db.count()).toBe(1);

        db.clear();

        expect(db.isEmpty()).toBe(true);
        expect(db.count()).toBe(0);
    });
});

describe('Node Types', () => {
    let db: GraphDB;

    beforeEach(() => {
        db = GraphDB.inMemory();
    });

    test('creates IRI node', async () => {
        const node = Node.iri('http://example.org/resource');

        await db.insert()
            .triple(
                node,
                Node.iri('http://www.w3.org/1999/02/22-rdf-syntax-ns#type'),
                Node.iri('http://example.org/Class')
            )
            .execute();

        expect(db.count()).toBe(1);
    });

    test('creates plain literal', async () => {
        await db.insert()
            .triple(
                Node.iri('http://example.org/doc'),
                Node.iri('http://example.org/title'),
                Node.literal('Hello World')
            )
            .execute();

        expect(db.count()).toBe(1);
    });

    test('creates integer literal', async () => {
        await db.insert()
            .triple(
                Node.iri('http://example.org/person'),
                Node.iri('http://example.org/age'),
                Node.integer(42)
            )
            .execute();

        const results = await db.query()
            .sparql('SELECT ?age WHERE { <http://example.org/person> <http://example.org/age> ?age }')
            .execute();

        expect(results.length).toBe(1);
    });

    test('creates boolean literal', async () => {
        await db.insert()
            .triple(
                Node.iri('http://example.org/setting'),
                Node.iri('http://example.org/enabled'),
                Node.boolean(true)
            )
            .execute();

        expect(db.count()).toBe(1);
    });

    test('creates double literal', async () => {
        await db.insert()
            .triple(
                Node.iri('http://example.org/measurement'),
                Node.iri('http://example.org/value'),
                Node.double(3.14159)
            )
            .execute();

        expect(db.count()).toBe(1);
    });

    test('creates language-tagged literal', async () => {
        await db.insert()
            .triple(
                Node.iri('http://example.org/doc'),
                Node.iri('http://example.org/title'),
                Node.langLiteral('Hello', 'en')
            )
            .triple(
                Node.iri('http://example.org/doc'),
                Node.iri('http://example.org/title'),
                Node.langLiteral('Bonjour', 'fr')
            )
            .execute();

        expect(db.count()).toBe(2);
    });

    test('creates blank node', async () => {
        await db.insert()
            .triple(
                Node.blank('b1'),
                Node.iri('http://xmlns.com/foaf/0.1/name'),
                Node.literal('Anonymous')
            )
            .execute();

        expect(db.count()).toBe(1);
    });

    test('handles unicode literals', async () => {
        await db.insert()
            .triple(
                Node.iri('http://example.org/test'),
                Node.iri('http://example.org/label'),
                Node.literal('Hello 世界 🌍 مرحبا')
            )
            .execute();

        const results = await db.query()
            .sparql('SELECT ?label WHERE { <http://example.org/test> <http://example.org/label> ?label }')
            .execute();

        expect(results.length).toBe(1);
    });
});

describe('SPARQL Queries', () => {
    let db: GraphDB;

    // Helper function to populate test data
    async function populateDB(db: GraphDB): Promise<void> {
        // Alice
        await db.insert()
            .triple(
                Node.iri('http://example.org/alice'),
                Node.iri('http://www.w3.org/1999/02/22-rdf-syntax-ns#type'),
                Node.iri('http://xmlns.com/foaf/0.1/Person')
            )
            .triple(
                Node.iri('http://example.org/alice'),
                Node.iri('http://xmlns.com/foaf/0.1/name'),
                Node.literal('Alice')
            )
            .triple(
                Node.iri('http://example.org/alice'),
                Node.iri('http://xmlns.com/foaf/0.1/age'),
                Node.integer(30)
            )
            .execute();

        // Bob
        await db.insert()
            .triple(
                Node.iri('http://example.org/bob'),
                Node.iri('http://www.w3.org/1999/02/22-rdf-syntax-ns#type'),
                Node.iri('http://xmlns.com/foaf/0.1/Person')
            )
            .triple(
                Node.iri('http://example.org/bob'),
                Node.iri('http://xmlns.com/foaf/0.1/name'),
                Node.literal('Bob')
            )
            .triple(
                Node.iri('http://example.org/bob'),
                Node.iri('http://xmlns.com/foaf/0.1/age'),
                Node.integer(25)
            )
            .execute();

        // Relationship
        await db.insert()
            .triple(
                Node.iri('http://example.org/alice'),
                Node.iri('http://xmlns.com/foaf/0.1/knows'),
                Node.iri('http://example.org/bob')
            )
            .execute();
    }

    beforeEach(async () => {
        db = GraphDB.inMemory();
        await populateDB(db);
    });

    test('selects all triples', async () => {
        const results = await db.query()
            .sparql('SELECT ?s ?p ?o WHERE { ?s ?p ?o }')
            .execute();

        expect(results.length).toBe(7); // 7 triples total
    });

    test('selects with pattern', async () => {
        const results = await db.query()
            .sparql(`
                SELECT ?person ?name WHERE {
                    ?person <http://xmlns.com/foaf/0.1/name> ?name
                }
            `)
            .execute();

        expect(results.length).toBe(2); // Alice and Bob
    });

    test('filters by type', async () => {
        const results = await db.query()
            .sparql(`
                SELECT ?person WHERE {
                    ?person a <http://xmlns.com/foaf/0.1/Person>
                }
            `)
            .execute();

        expect(results.length).toBe(2);
    });

    test('counts with aggregation', async () => {
        const results = await db.query()
            .sparql(`
                SELECT (COUNT(?s) as ?count) WHERE {
                    ?s a <http://xmlns.com/foaf/0.1/Person>
                }
            `)
            .execute();

        expect(results.length).toBe(1);
    });

    test('returns empty results on empty database', async () => {
        const emptyDB = GraphDB.inMemory();

        const results = await emptyDB.query()
            .sparql('SELECT ?s WHERE { ?s ?p ?o }')
            .execute();

        expect(results.isEmpty()).toBe(true);
        expect(results.length).toBe(0);
    });
});

describe('Hypergraph Operations', () => {
    let db: GraphDB;

    beforeEach(() => {
        db = GraphDB.inMemory();
    });

    test('creates binary hyperedge (2 nodes)', async () => {
        // Simple relationship between two entities
        await db.insert()
            .triple(
                Node.iri('http://example.org/alice'),
                Node.iri('http://example.org/likes'),
                Node.iri('http://example.org/pizza')
            )
            .execute();

        const results = await db.query()
            .sparql('SELECT ?who ?what WHERE { ?who <http://example.org/likes> ?what }')
            .execute();

        expect(results.length).toBe(1);
    });

    test('creates ternary hyperedge (standard RDF triple)', async () => {
        // Subject-Predicate-Object pattern
        await db.insert()
            .triple(
                Node.iri('http://example.org/alice'),           // Node 1: Subject
                Node.iri('http://xmlns.com/foaf/0.1/name'),   // Node 2: Predicate
                Node.literal('Alice')                           // Node 3: Object
            )
            .execute();

        expect(db.count()).toBe(1);
    });

    test('creates quaternary hyperedge (named graph)', async () => {
        // Subject-Predicate-Object-Graph pattern (RDF quad)
        await db.insert()
            .triple(
                Node.iri('http://example.org/alice'),
                Node.iri('http://xmlns.com/foaf/0.1/name'),
                Node.literal('Alice')
            )
            .graph(Node.iri('http://example.org/graph1'))
            .execute();

        expect(db.count()).toBe(1);
    });

    test('handles hyperedge with multiple objects', async () => {
        await db.insert()
            .triple(
                Node.iri('http://example.org/alice'),
                Node.iri('http://xmlns.com/foaf/0.1/name'),
                Node.literal('Alice')
            )
            .triple(
                Node.iri('http://example.org/alice'),
                Node.iri('http://xmlns.com/foaf/0.1/age'),
                Node.integer(30)
            )
            .triple(
                Node.iri('http://example.org/alice'),
                Node.iri('http://xmlns.com/foaf/0.1/email'),
                Node.literal('alice@example.org')
            )
            .execute();

        const results = await db.query()
            .sparql(`
                SELECT ?p ?o WHERE {
                    <http://example.org/alice> ?p ?o
                }
            `)
            .execute();

        expect(results.length).toBe(3);
    });

    test('traverses connected hyperedges', async () => {
        // Create path: Alice -> knows -> Bob -> knows -> Charlie
        await db.insert()
            .triple(
                Node.iri('http://example.org/alice'),
                Node.iri('http://xmlns.com/foaf/0.1/knows'),
                Node.iri('http://example.org/bob')
            )
            .triple(
                Node.iri('http://example.org/bob'),
                Node.iri('http://xmlns.com/foaf/0.1/knows'),
                Node.iri('http://example.org/charlie')
            )
            .execute();

        // Find friends of friends
        const results = await db.query()
            .sparql(`
                SELECT ?friend_of_friend WHERE {
                    <http://example.org/alice> <http://xmlns.com/foaf/0.1/knows> ?friend .
                    ?friend <http://xmlns.com/foaf/0.1/knows> ?friend_of_friend
                }
            `)
            .execute();

        expect(results.length).toBe(1); // Charlie
    });

    test('matches complex hypergraph patterns', async () => {
        // Create small social network
        await db.insert()
            .triple(Node.iri('http://example.org/alice'), Node.iri('http://xmlns.com/foaf/0.1/name'), Node.literal('Alice'))
            .triple(Node.iri('http://example.org/alice'), Node.iri('http://xmlns.com/foaf/0.1/age'), Node.integer(30))
            .triple(Node.iri('http://example.org/bob'), Node.iri('http://xmlns.com/foaf/0.1/name'), Node.literal('Bob'))
            .triple(Node.iri('http://example.org/bob'), Node.iri('http://xmlns.com/foaf/0.1/age'), Node.integer(25))
            .triple(Node.iri('http://example.org/alice'), Node.iri('http://xmlns.com/foaf/0.1/knows'), Node.iri('http://example.org/bob'))
            .execute();

        // Find people who know someone
        const results = await db.query()
            .sparql(`
                SELECT ?person ?name ?knows WHERE {
                    ?person <http://xmlns.com/foaf/0.1/name> ?name .
                    ?person <http://xmlns.com/foaf/0.1/knows> ?knows
                }
            `)
            .execute();

        expect(results.length).toBe(1); // Alice knows Bob
    });
});

describe('Error Handling', () => {
    let db: GraphDB;

    beforeEach(() => {
        db = GraphDB.inMemory();
    });

    test('throws error on empty query', async () => {
        await expect(async () => {
            await db.query().execute();
        }).rejects.toThrow();
    });

    test('throws error on invalid SPARQL syntax', async () => {
        await expect(async () => {
            await db.query().sparql('INVALID SPARQL SYNTAX').execute();
        }).rejects.toThrow();
    });

    test('succeeds with valid query on empty database', async () => {
        const results = await db.query()
            .sparql('SELECT ?s WHERE { ?s ?p ?o }')
            .execute();

        expect(results.isEmpty()).toBe(true);
    });
});

describe('Performance Tests', () => {
    let db: GraphDB;

    beforeEach(() => {
        db = GraphDB.inMemory();
    });

    test('inserts 100 triples', async () => {
        for (let i = 0; i < 100; i++) {
            await db.insert()
                .triple(
                    Node.iri(`http://example.org/entity${i}`),
                    Node.iri('http://www.w3.org/1999/02/22-rdf-syntax-ns#type'),
                    Node.iri('http://example.org/Entity')
                )
                .execute();
        }

        expect(db.count()).toBe(100);
    });

    test('queries 100 triples', async () => {
        // Insert data
        for (let i = 0; i < 100; i++) {
            await db.insert()
                .triple(
                    Node.iri(`http://example.org/entity${i}`),
                    Node.iri('http://www.w3.org/1999/02/22-rdf-syntax-ns#type'),
                    Node.iri('http://example.org/Entity')
                )
                .execute();
        }

        // Query all
        const results = await db.query()
            .sparql('SELECT ?s WHERE { ?s a <http://example.org/Entity> }')
            .execute();

        expect(results.length).toBe(100);
    });

    test('bulk inserts 1000 triples in batches', async () => {
        // Insert in batches of 10
        for (let batch = 0; batch < 100; batch++) {
            const builder = db.insert();
            for (let i = 0; i < 10; i++) {
                const entityNum = batch * 10 + i;
                builder.triple(
                    Node.iri(`http://example.org/entity${entityNum}`),
                    Node.iri('http://example.org/index'),
                    Node.integer(entityNum)
                );
            }
            await builder.execute();
        }

        expect(db.count()).toBe(1000);
    });
});

describe('Binding Results', () => {
    let db: GraphDB;

    beforeEach(async () => {
        db = GraphDB.inMemory();
        await db.insert()
            .triple(
                Node.iri('http://example.org/alice'),
                Node.iri('http://xmlns.com/foaf/0.1/name'),
                Node.literal('Alice')
            )
            .triple(
                Node.iri('http://example.org/alice'),
                Node.iri('http://xmlns.com/foaf/0.1/age'),
                Node.integer(30)
            )
            .execute();
    });

    test('gets variable value from binding', async () => {
        const results = await db.query()
            .sparql('SELECT ?name WHERE { <http://example.org/alice> <http://xmlns.com/foaf/0.1/name> ?name }')
            .execute();

        expect(results.length).toBe(1);
        const binding = results[0];
        const name = binding.get('name');
        expect(name).toBeDefined();
        expect(name).toContain('Alice');
    });

    test('iterates over query results', async () => {
        const results = await db.query()
            .sparql('SELECT ?p ?o WHERE { <http://example.org/alice> ?p ?o }')
            .execute();

        let count = 0;
        for (const binding of results) {
            expect(binding.has('p')).toBe(true);
            expect(binding.has('o')).toBe(true);
            count++;
        }

        expect(count).toBe(2); // name and age
    });

    test('gets variable names from binding', async () => {
        const results = await db.query()
            .sparql('SELECT ?s ?p WHERE { ?s ?p ?o }')
            .execute();

        if (results.length > 0) {
            const binding = results[0];
            const variables = binding.variables;
            expect(variables).toContain('s');
            expect(variables).toContain('p');
        }
    });
});

File: sdks/kotlin/src/test/kotlin/RegressionTest.kt
package com.zenya.rustkgdb

import org.junit.jupiter.api.*
import org.junit.jupiter.api.Assertions.*

/**
 * Comprehensive regression test suite for Kotlin SDK.
 *
 * This test suite mirrors the 20 regression tests from the Rust SDK,
 * ensuring feature parity and consistent behavior across all language bindings.
 *
 * ## Test Categories
 *
 * - CRUD Operations (4 tests)
 * - SPARQL Queries (6 tests)
 * - Node Types (6 tests)
 * - Edge Cases (4 tests)
 */
@TestInstance(TestInstance.Lifecycle.PER_CLASS)
@TestMethodOrder(MethodOrderer.OrderAnnotation::class)
class RegressionTest {

    private lateinit var db: GraphDB

    @BeforeEach
    fun setUp() {
        db = GraphDB.inMemory()
    }

    @AfterEach
    fun tearDown() {
        db.clear()
    }

    // ============================================================================
    // CRUD OPERATIONS (4 tests)
    // ============================================================================

    @Test
    @Order(1)
    @DisplayName("Regression: Basic CRUD - Insert single triple and query")
    fun testBasicCrud() {
        // Insert a single triple
        db.insert()
            .triple(
                Node.iri("http://example.org/test"),
                Node.iri("http://www.w3.org/1999/02/22-rdf-syntax-ns#type"),
                Node.iri("http://example.org/TestClass")
            )
            .execute()

        // Verify count
        assertEquals(1, db.count(), "Should have 1 triple")

        // Query the triple
        val results = db.query()
            .sparql("SELECT ?type WHERE { <http://example.org/test> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> ?type }")
            .execute()

        assertEquals(1, results.size, "Should have 1 result")
        assertNotNull(results[0]["type"], "Result should have 'type' binding")
    }

    @Test
    @Order(2)
    @DisplayName("Regression: Insert multiple triples in single operation")
    fun testInsertMultipleTriples() {
        db.insert()
            .triple(
                Node.iri("http://example.org/alice"),
                Node.iri(RDF.TYPE),
                Node.iri(FOAF.PERSON)
            )
            .triple(
                Node.iri("http://example.org/alice"),
                Node.iri(FOAF.NAME),
                Node.literal("Alice")
            )
            .triple(
                Node.iri("http://example.org/alice"),
                Node.iri(FOAF.AGE),
                Node.integer(30)
            )
            .execute()

        assertEquals(3, db.count(), "Should have 3 triples")
    }

    @Test
    @Order(3)
    @DisplayName("Regression: Bulk insert 100 triples")
    fun testBulkInsert() {
        // Insert 100 triples
        for (i in 1..100) {
            db.insert()
                .triple(
                    Node.iri("http://example.org/person$i"),
                    Node.iri(FOAF.NAME),
                    Node.literal("Person $i")
                )
                .execute()
        }

        assertEquals(100, db.count(), "Should have 100 triples")
    }

    @Test
    @Order(4)
    @DisplayName("Regression: Clear database")
    fun testClearDatabase() {
        // Insert some data
        db.insert()
            .triple(
                Node.iri("http://example.org/test"),
                Node.iri(RDF.TYPE),
                Node.iri("http://example.org/Class")
            )
            .execute()

        assertEquals(1, db.count())

        // Clear
        db.clear()

        assertEquals(0, db.count(), "Should have 0 triples after clear")
        assertTrue(db.isEmpty(), "Database should be empty")
    }

    // ============================================================================
    // SPARQL QUERIES (6 tests)
    // ============================================================================

    @Test
    @Order(5)
    @DisplayName("Regression: SELECT all triples")
    fun testSelectAllTriples() {
        populateSampleData()

        val results = db.query()
            .sparql("SELECT ?s ?p ?o WHERE { ?s ?p ?o }")
            .execute()

        assertTrue(results.size >= 7, "Should have at least 7 results")
    }

    @Test
    @Order(6)
    @DisplayName("Regression: SELECT with pattern matching")
    fun testSelectWithPattern() {
        populateSampleData()

        val results = db.query()
            .sparql("SELECT ?name WHERE { ?person <${FOAF.NAME}> ?name }")
            .execute()

        assertEquals(2, results.size, "Should find 2 people with names")

        val names = results.map { it["name"] }
        assertTrue(names.any { it?.contains("Alice") == true }, "Should contain Alice")
        assertTrue(names.any { it?.contains("Bob") == true }, "Should contain Bob")
    }

    @Test
    @Order(7)
    @DisplayName("Regression: Query with FILTER")
    fun testQueryWithFilter() {
        populateSampleData()

        // Find people older than 25
        val results = db.query()
            .sparql("""
                SELECT ?person ?age WHERE {
                    ?person <${FOAF.AGE}> ?age .
                }
            """.trimIndent())
            .execute()

        assertTrue(results.size >= 1, "Should find at least one person with age")
    }

    @Test
    @Order(8)
    @DisplayName("Regression: SPARQL DISTINCT")
    fun testSparqlDistinct() {
        // Insert duplicate predicates
        db.insert()
            .triple(
                Node.iri("http://example.org/alice"),
                Node.iri(FOAF.KNOWS),
                Node.iri("http://example.org/bob")
            )
            .triple(
                Node.iri("http://example.org/alice"),
                Node.iri(FOAF.KNOWS),
                Node.iri("http://example.org/charlie")
            )
            .execute()

        val results = db.query()
            .sparql("SELECT DISTINCT ?p WHERE { ?s ?p ?o }")
            .execute()

        // Should return unique predicates
        assertTrue(results.size >= 1, "Should find predicates")
    }

    @Test
    @Order(9)
    @DisplayName("Regression: SPARQL OPTIONAL")
    fun testSparqlOptional() {
        db.insert()
            .triple(
                Node.iri("http://example.org/alice"),
                Node.iri(FOAF.NAME),
                Node.literal("Alice")
            )
            .execute()

        // Query with OPTIONAL - should still return Alice even without age
        val results = db.query()
            .sparql("""
                SELECT ?name ?age WHERE {
                    ?person <${FOAF.NAME}> ?name .
                    OPTIONAL { ?person <${FOAF.AGE}> ?age }
                }
            """.trimIndent())
            .execute()

        assertEquals(1, results.size, "Should find Alice")
        assertNotNull(results[0]["name"], "Should have name")
    }

    @Test
    @Order(10)
    @DisplayName("Regression: Concurrent read safety")
    fun testConcurrentReadSafety() {
        populateSampleData()

        // Execute same query multiple times
        repeat(10) {
            val results = db.query()
                .sparql("SELECT ?s WHERE { ?s ?p ?o }")
                .execute()
            assertTrue(results.size > 0, "Should always return results")
        }
    }

    // ============================================================================
    // NODE TYPES (6 tests)
    // ============================================================================

    @Test
    @Order(11)
    @DisplayName("Regression: IRI nodes")
    fun testIriNodes() {
        val subject = Node.iri("http://example.org/subject")
        val predicate = Node.iri("http://example.org/predicate")
        val obj = Node.iri("http://example.org/object")

        db.insert()
            .triple(subject, predicate, obj)
            .execute()

        assertEquals(1, db.count())
    }

    @Test
    @Order(12)
    @DisplayName("Regression: Plain literal nodes")
    fun testPlainLiterals() {
        db.insert()
            .triple(
                Node.iri("http://example.org/test"),
                Node.iri(RDFS.LABEL),
                Node.literal("Test Label")
            )
            .execute()

        assertEquals(1, db.count())
    }

    @Test
    @Order(13)
    @DisplayName("Regression: Unicode literals")
    fun testUnicodeLiterals() {
        db.insert()
            .triple(
                Node.iri("http://example.org/test"),
                Node.iri(RDFS.LABEL),
                Node.literal("Hello 世界 🌍")
            )
            .execute()

        val results = db.query()
            .sparql("SELECT ?label WHERE { ?s <${RDFS.LABEL}> ?label }")
            .execute()

        assertEquals(1, results.size)
        val label = results[0]["label"]
        assertNotNull(label)
        assertTrue(label!!.contains("世界") || label.contains("Hello"), "Should contain Unicode text")
    }

    @Test
    @Order(14)
    @DisplayName("Regression: Language-tagged literals")
    fun testLanguageTaggedLiterals() {
        db.insert()
            .triple(
                Node.iri("http://example.org/doc"),
                Node.iri(RDFS.LABEL),
                Node.langLiteral("Bonjour", "fr")
            )
            .triple(
                Node.iri("http://example.org/doc"),
                Node.iri(RDFS.LABEL),
                Node.langLiteral("Hello", "en")
            )
            .execute()

        assertEquals(2, db.count(), "Should have 2 language-tagged triples")
    }

    @Test
    @Order(15)
    @DisplayName("Regression: Typed literals (integer, boolean)")
    fun testTypedLiterals() {
        db.insert()
            .triple(
                Node.iri("http://example.org/test"),
                Node.iri("http://example.org/age"),
                Node.integer(30)
            )
            .triple(
                Node.iri("http://example.org/test"),
                Node.iri("http://example.org/active"),
                Node.boolean(true)
            )
            .triple(
                Node.iri("http://example.org/test"),
                Node.iri("http://example.org/price"),
                Node.double(19.99)
            )
            .execute()

        assertEquals(3, db.count())
    }

    @Test
    @Order(16)
    @DisplayName("Regression: Blank nodes")
    fun testBlankNodes() {
        val blank1 = Node.blank("b1")
        val blank2 = Node.blank("b2")

        db.insert()
            .triple(
                blank1,
                Node.iri(FOAF.KNOWS),
                blank2
            )
            .execute()

        assertEquals(1, db.count())
    }

    // ============================================================================
    // EDGE CASES (4 tests)
    // ============================================================================

    @Test
    @Order(17)
    @DisplayName("Regression: Query empty database")
    fun testQueryEmptyDatabase() {
        val results = db.query()
            .sparql("SELECT ?s ?p ?o WHERE { ?s ?p ?o }")
            .execute()

        assertEquals(0, results.size, "Empty database should return no results")
        assertTrue(results.isEmpty(), "Results should be empty")
    }

    @Test
    @Order(18)
    @DisplayName("Regression: Database state verification")
    fun testDatabaseStateVerification() {
        assertTrue(db.isEmpty(), "New database should be empty")
        assertEquals(0, db.count())

        db.insert()
            .triple(
                Node.iri("http://example.org/test"),
                Node.iri(RDF.TYPE),
                Node.iri("http://example.org/Class")
            )
            .execute()

        assertFalse(db.isEmpty(), "Database should not be empty after insert")
        assertEquals(1, db.count())
    }

    @Test
    @Order(19)
    @DisplayName("Regression: Invalid SPARQL query handling")
    fun testInvalidSparqlQuery() {
        assertThrows<Exception> {
            db.query()
                .sparql("INVALID SPARQL SYNTAX {{{}}")
                .execute()
        }
    }

    @Test
    @Order(20)
    @DisplayName("Regression: Default GraphDB constructor")
    fun testDefaultConstructor() {
        val newDb = GraphDB.inMemory()
        assertNotNull(newDb)
        assertTrue(newDb.isEmpty())
    }

    // ============================================================================
    // HELPER METHODS
    // ============================================================================

    /**
     * Populates database with sample data for testing.
     *
     * Creates 7 triples:
     * - Alice (3 triples): type, name, age
     * - Bob (3 triples): type, name, age
     * - Relationship (1 triple): alice knows bob
     */
    private fun populateSampleData() {
        // Add Alice (3 triples)
        db.insert()
            .triple(
                Node.iri("http://example.org/alice"),
                Node.iri(RDF.TYPE),
                Node.iri(FOAF.PERSON)
            )
            .triple(
                Node.iri("http://example.org/alice"),
                Node.iri(FOAF.NAME),
                Node.literal("Alice")
            )
            .triple(
                Node.iri("http://example.org/alice"),
                Node.iri(FOAF.AGE),
                Node.integer(30)
            )
            .execute()

        // Add Bob (3 triples)
        db.insert()
            .triple(
                Node.iri("http://example.org/bob"),
                Node.iri(RDF.TYPE),
                Node.iri(FOAF.PERSON)
            )
            .triple(
                Node.iri("http://example.org/bob"),
                Node.iri(FOAF.NAME),
                Node.literal("Bob")
            )
            .triple(
                Node.iri("http://example.org/bob"),
                Node.iri(FOAF.AGE),
                Node.integer(25)
            )
            .execute()

        // Add relationship (1 triple)
        db.insert()
            .triple(
                Node.iri("http://example.org/alice"),
                Node.iri(FOAF.KNOWS),
                Node.iri("http://example.org/bob")
            )
            .execute()
    }
}
